src/configobject.cpp

```diff
@@ -274,60 +167,39 @@ template <class ValueType> bool ConfigObject<ValueType>::Parse()
     return true;
 }
 
-template <class ValueType> void ConfigObject<ValueType>::clear()
-{
-    //Delete the pointers, because that's what we did before we
-    //purged Mixxx of Qt3 code. -- Albert, June 18th 2010 (at 30,000 ft)
-    for (int i = 0; i < m_list.count(); i++)
-        delete m_list[i];
-
-    // This shouldn't be done, since objects might have references to
-    // members of list. Instead all member values should be set to some
-    // null value.
-    m_list.clear();
-
-}
-
-template <class ValueType> void ConfigObject<ValueType>::reopen(QString file)
-{
+template <class ValueType> void ConfigObject<ValueType>::reopen(QString file) {
     m_filename = file;
-    Parse();
+    parse();
 }
 
-template <class ValueType> void ConfigObject<ValueType>::Save()
-{
+template <class ValueType> void ConfigObject<ValueType>::Save() {
+    QMutexLocker lock(&m_valueHashMutex);
     QFile file(m_filename);
     if (!QDir(QFileInfo(file).absolutePath()).exists()) {
         QDir().mkpath(QFileInfo(file).absolutePath());
     }
-    if (!file.open(QIODevice::WriteOnly | QIODevice::Text))
-    {
+    if (!file.open(QIODevice::WriteOnly | QIODevice::Text)) {
         qDebug() << "Could not write file" << m_filename << ", don't worry.";
         return;
-    }
-    else
-    {
+    } else {
         QTextStream stream(&file);
         stream.setCodec("UTF-8");
 
         QString grp = "";
 
-        QListIterator<ConfigOption<ValueType>* > iterator(m_list);
-        ConfigOption<ValueType>* it;
-        while (iterator.hasNext())
-        {
-            it = iterator.next();
-//            qDebug() << "group:" << it->key->group << "item" << it->key->item << "val" << it->val->value;
-            if (it->key->group != grp)
-            {
-                grp = it->key->group;
-                stream << "\n" << it->key->group << "\n";
+        typename QHash<ConfigKey, ValueType>::iterator i;
+        for (i = m_valueHash.begin(); i != m_valueHash.end(); ++i) {
```

Comment:
This won't necessarily group same-group keys together since qHash(ConfigKey) can spread them in different hash buckets. A QMap may be better because its iteration order is sorted by its keys. Or we could group the ConfigKeys by group when saving.